### PR TITLE
Patch 1 for program from program creation

### PR DESCRIPTION
**Pull request by SabaunT**
Creating program from program implementation is on `st-create-program-from-program-iter1` branch and quite big. To make it easy for review, I decided to divide it to separate PRs.

This is the first one.

**Release notes**: a new sys-call for sending init messages in program is introduced.

@gear-tech/dev

### CODE REVIEW

File: core-backend/common/src/funcs.rs [hunk 1]
```diff
@@ -255,6 +255,36 @@ pub fn send_push<E: Ext>(ext: LaterExt<E>) -> impl Fn(i32, i32, i32) -> Result<(
...
 
+pub fn create_program<E: Ext>(
+    ext: LaterExt<E>,
+) -> impl Fn(i32, i32, i32, i32, i32, i64, i32, i32) -> Result<(), &'static str> {
+    move |code_hash_ptr: i32,
+          salt_ptr: i32,
+          salt_len: i32,
+          payload_ptr: i32,
+          payload_len: i32,
+          gas_limit: i64,
```
Comment (shamilsan) [inline core-backend/common/src/funcs.rs:266]: We are to split this method into two: with `gas_limit` and without. Maybe later.

Comment (SabaunT) [reply core-backend/common/src/funcs.rs:266]: Shall I leave the current one?

Comment (breathx) [reply core-backend/common/src/funcs.rs:266]: I think that's not the problem. We will gone out of gas limits only after #715, so it may be done later

Comment (SabaunT) [reply core-backend/common/src/funcs.rs:266]: I guess, resolve this

File: core-backend/sandbox/src/env.rs [hunk 2]
```diff
@@ -142,6 +142,7 @@ where
         env_builder.add_host_func("env", "gr_wait", funcs::wait);
         env_builder.add_host_func("env", "gr_wake", funcs::wake);
         env_builder.add_host_func("env", "gas", funcs::gas);
+        env_builder.add_host_func("env", "gr_create_program", funcs::create_program);
```
Comment (shamilsan) [inline core-backend/sandbox/src/env.rs:145]: We tried to sort syscalls alphabetically.

File: core-backend/common/src/funcs.rs (hunk 1 above)
Comment (shamilsan) [reply core-backend/common/src/funcs.rs:266]: We can rename it to `create_program_wgas`.

Comment (SabaunT) [reply core-backend/common/src/funcs.rs:266]: Maybe better later create `create_program_gas_less`?
Or default is create without gas?

File: core-processor/src/ext.rs [hunk 3]
```diff
@@ -443,4 +454,33 @@ impl EnvExt for Ext {
...
+            let mut data = Vec::with_capacity(code_hash.inner().len() + packet.salt.len());
+            code_hash.encode_to(&mut data);
+            packet.salt.encode_to(&mut data);
+            ProgramId::from_slice(blake2_rfc::blake2b::blake2b(32, &[], &data).as_bytes())
+        };
+
+        let entry = self
+            .program_candidates_data
+            .entry(code_hash)
+            .or_insert(Vec::new());
```
Comment (breathx) [inline core-processor/src/ext.rs:470]: or_insert_default

File: core-backend/sandbox/src/env.rs (hunk 2 above)
Comment (SabaunT) [reply core-backend/sandbox/src/env.rs]: done

File: core-backend/common/src/funcs.rs (hunk 1 above)
Comment (shamilsan) [reply core-backend/common/src/funcs.rs:266]: See `send_wgas` etc in #715

File: core-processor/src/ext.rs [hunk 4]
```diff
@@ -443,4 +454,33 @@ impl EnvExt for Ext {
...
+            return self.return_and_store_err(Err("Duplicate init message for the same id"));
+        }
+
+        // Send a message for program creation
+        let (new_prog_id, init_msg_id) = self
+            .message_context
+            .send_init_program(new_program_id, packet);
+
+        // Save a program candidate for this run
+        entry.push((new_prog_id, init_msg_id));
```
Comment (shamilsan) [inline core-processor/src/ext.rs:482]: How do you use this `entry` later?

File: core-backend/common/src/funcs.rs (hunk 1 above)
Comment (SabaunT) [reply core-backend/common/src/funcs.rs:266]: Done

File: gstd/src/msg/basic.rs [hunk 5]
```diff
@@ -474,3 +474,20 @@ pub fn source() -> ActorId {
...
 }
+
+pub fn create_program_wgas<T1: AsRef<[u8]>, T2: AsRef<[u8]>>(
+    code_hash: CodeHash,
+    salt: T1,
+    payload: T2,
+    gas_limit: u64,
+    value: u128,
+) -> ActorId {
+    gcore::msg::create_program_wgas(
```
Comment (breathx) [inline gstd/src/msg/basic.rs:485]: considering about where should this method be

is `msg` module is suitable enough for that logic?

File: core/src/message.rs [hunk 6]
```diff
@@ -595,6 +691,8 @@ pub trait MessageIdGenerator {
 pub struct MessageState {
     /// Collection of outgoing messages generated.
     pub outgoing: Vec<OutgoingMessage>,
+    /// Collection of init messages for new programs generated.
```
Comment (breathx) [inline core/src/message.rs:694]: Should be saved across wait-wakes logic via setting in `PayloadStore`, see #684

File: core-processor/src/ext.rs [hunk 7]
```diff
@@ -182,17 +189,21 @@ impl From<Ext> for ExtInfo {
...
             accessed_pages,
+            init_messages,
             outgoing,
             reply,
             awakening,
             nonce,
             payload_store: Some(store),
             trap_explanation,
             exit_argument: ext.exit_argument,
+            program_candidates_data,
```
Comment (breathx) [inline core-processor/src/ext.rs:206]: Minor refactoring

```suggestion
        ExtInfo {
            gas_amount,
            pages: ext.memory_context.allocations().clone(),
            accessed_pages,
            init_messages,
            outgoing,
            reply,
            awakening,
            nonce,
            payload_store: Some(store),
            trap_explanation: ext.error_explanation,
            exit_argument: ext.exit_argument,
            program_candidates_data: ext.program_candidates_data,
```

**Review by breathx — CHANGES_REQUESTED** [review]
Once comment about saving context will be resolved, It will be ready to be merged.

Great stuff! 👍

File: gstd/src/msg/basic.rs (hunk 5 above)
Comment (SabaunT) [reply gstd/src/msg/basic.rs:485]: well, basically, we send `init` message, so seems that should in messages module. What do you think?

**Review by shamilsan — APPROVED** [review]

File: core-processor/src/ext.rs (hunk 7 above)
Comment (SabaunT) [reply core-processor/src/ext.rs]: done

File: core/src/message.rs (hunk 6 above)
Comment (SabaunT) [reply core/src/message.rs:684]: done

File: gstd/src/msg/basic.rs (hunk 5 above)
Comment (SabaunT) [reply gstd/src/msg/basic.rs]: decided to move to `prog`

File: core-processor/src/ext.rs (hunk 3 above)
Comment (SabaunT) [reply core-processor/src/ext.rs]: done

File: gstd/src/prog/mod.rs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Program creation module
+
+use crate::{ActorId, CodeHash};
+use codec::Encode;
+
+pub fn create_program_wgas<T1: Encode, T2: Encode>(
```
Comment (breathx) [inline gstd/src/prog/mod.rs:24]: Why not `AsRef<[u8]>`, but `Encode` ?

Comment (SabaunT) [reply gstd/src/prog/mod.rs]: @shamilsan told me that `gstd` API has the major difference from `gcore` in using `Encode` and `Decode`. So I applied that rule here.

**Pull request merged**